src/app/services/contracts.service.ts

```diff
@@ -0,0 +1,156 @@
+import {HttpClient} from '@angular/common/http';
+import {NGXLogger} from 'ngx-logger';
+import {Observable} from 'rxjs';
+import {environment} from '../../environments/environment';
+import {catchError, filter, map} from 'rxjs/operators';
+import {SnackService} from './snack.service';
+import {StaticValues} from '../static/static-values';
+import {Addresses} from '../static/addresses';
+
+interface Result<T> {
+    data: T | Array<T>;
+}
+
+// eslint-disable-next-line
+export interface IsContract{}
+
+export class Contract implements IsContract {
+    id: number;
+    address: string;
+    name: string;
+    prettyName() {
+        return this.name
+            ?.replace('SUSHI_', '')
+            ?.replace('ONEINCH_', '')
+            ?.replace('UNI_LP_', '')
+            ?.replace('UNI_', '')
+            ?.replace('ST_', '')
+            ?.replace('_HODL', '_H')
+            ?.replace('HODL', 'SUSHI_HODL');
+    }
+    created: number;
+    type: number;
+}
+
+export class Vault implements IsContract {
+    id: number;
+    contract: Contract;
+    updatedBlock: number;
+    controller: Contract;
+    governance: Contract;
+    strategy: Contract;
+    underlying: Contract;
+    name: string;
+    symbol: string;
+    decimals: number;
+    underlyingUnit: number;
+    isStableCoin() {
+        return this.underlying?.type === 4 && !this.contract?.name?.match(/CRV/);
+    }
+    isCRV() {
+        return this.underlying?.type === 4 && this.contract?.name?.match(/CRV/);
+    }
+    isUniLP() {
+        return this.underlying?.type === 2 && this.underlying?.name?.match(/^UNI_LP/);
+    }
+    isOtherLP() {
+        return this.underlying?.type === 2 && !this.underlying?.name?.match(/^UNI_LP/);
+    }
+}
+
+export class Pool implements IsContract {
+    id: number;
+    contract: Contract;
+    updatedBlock: number;
+    controller: Contract;
+    governance: Contract;
+    owner: Contract;
+    lpToken: Contract;
+    rewardToken: Contract;
+}
+
+export class Token implements IsContract {
+    id: number;
+    contract: Contract;
+    updatedBlock: number;
+    name: string;
+    symbol: string;
+    decimals: number;
+}
+
+export class Pair implements IsContract {
+    id: number;
+    contract: Contract;
+    updatedBlock: number;
+    keyToken: Token;
+    type: number;
+    decimals: number;
+    token0: Contract;
+    token1: Contract;
+}
+
+interface Service<T extends IsContract> {
+    getContract(type: new () => T, name: string): Observable<T>;
+    getContracts(type: new () => T): Observable<T[]>;
+}
+
+const TYPE_PATHS = new Map<IsContract, string>([[Vault, 'vault'], [Pool, 'pool'], [Token, 'token'], [Pair, 'unipair']]);
+
+@Injectable({
+    providedIn: 'root'
+})
+export class ContractsService<T> implements Service<T> {
+
+    private urlPrefix = 'contracts';
+
+    constructor(private http: HttpClient, private snackService: SnackService) {
+
+    }
+
+    getContract(type: new () => T, name: string): Observable<T> {
```

Comment:
use generic for function instead of class
otherwise, we need multiple implementations for basic calls

Reply:
👍